### DISCUSSION TITLE
Convert tailwind classname to Javascript object

### DISCUSSION THREAD

**Discussion by stevenmunro** [tailwindlabs/tailwindcss#9488] [Help] [answered]
I'm writing something to help me style PDFs using react-pdf, and I want to use tailwindcss. All I need to do is have something that takes a tailwind class name like;

`text-3xl`

and give me

`{fontSize: 1.875rem, lineHeight: 2.25rem}`

I'd like to have this function scour my own config using `resolveConfig`

At the moment, I have something working using *https://github.com/Devzstudio/tailwind_to_css* and running the output through a camelCase converter function. This tailwind_to_css library uses a cheatsheet *https://raw.githubusercontent.com/Devzstudio/tailwind_to_css/main/cheatsheet.ts* so it doesn't incorporate any custom styles or fonts at all. This approach feels wobbly and I'm hoping that someone has a more elegant solution.

**Comment by wongjn** [comment]
I think Tailwind already uses some JavaScript object data type, since it uses [postcss-js](https://github.com/postcss/postcss-js) under the hood, for example:
https://github.com/tailwindlabs/tailwindcss/blob/a34bd62bb6a82886c1a03611b1b85ccae37a5f7b/src/util/parseObjectStyles.js#L1-L19

You might be able to replicate or "intercept" the CSS code generation before it gets put back into a CSS string by PostCSS. This would then allow you to get the CSS properties for a class.

An alternative solution could be that you use postcss-js after the CSS generation and then [parse it back into object syntax](https://github.com/postcss/postcss-js#compile-css-to-css-in-js).

**Comment by stevenmunro** [comment, marked as answer]
Hi @wongjn. Thank you for your quick response. Your advice helped me realise what a complex beast Tailwindcss is, used it for years, love it, but didn't realise just how much is going on under the hood.

Rather than fight with borrowing functions out of the core project, I decided the best way to remain compatible with future releases of Tailwindcss is to stick to their very own Tailwindcss CLI tool.

The CSS output contains everything I need including custom fonts defined in `tailwind.config.js`.

Then, I run the CSS file through a parser `(css.parse)` *https://github.com/reworkcss/css* to obtain a lovely AST object.

Then it was just a simple case of iterating over the rules, camel casing and removing backslashes from the selectors, the iterating over the declarations, putting them into nice little objects.

For anyone looking to do this themselves. This is what an AST object looks like.

``` js
{
  "ast": {
    "type": "stylesheet",
    "stylesheet": {
      "source": "pdf-one.css",
      "rules": [
        {
          "type": "rule",
          "selectors": [
            ".inset-3\\/4"
          ],
          "declarations": [
            {
              "type": "declaration",
              "property": "top",
              "value": "75%",
              "position": {
                "start": {
                  "line": 2,
                  "column": 3
                },
                "end": {
                  "line": 2,
                  "column": 11
                },
                "source": "pdf-one.css"
              }
            },
            {
              "type": "declaration",
              "property": "right",
              "value": "75%",
              "position": {
                "start": {
                  "line": 3,
                  "column": 3
                },
                "end": {
                  "line": 3,
                  "column": 13
                },
                "source": "pdf-one.css"
              }
            },
            {
              "type": "declaration",
              "property": "bottom",
              "value": "75%",
              "position": {
                "start": {
                  "line": 4,
                  "column": 3
                },
                "end": {
                  "line": 4,
                  "column": 14
                },
                "source": "pdf-one.css"
              }
            },
            {
              "type": "declaration",
              "property": "left",
              "value": "75%",
              "position": {
                "start": {
                  "line": 5,
                  "column": 3
                },
                "end": {
                  "line": 5,
                  "column": 12
                },
                "source": "pdf-one.css"
              }
            }
          ],
          "position": {
            "start": {
              "line": 1,
              "column": 1
            },
            "end": {
              "line": 6,
              "column": 2
            },
            "source": "pdf-one.css"
          }
        }
      ],
      "parsingErrors": []
    }
  }
}
```

and after iterating over the tree, this is the output I want;

``` js
{
  "inset-3/4": {
    "top": "75%",
    "right": "75%",
    "bottom": "75%",
    "left": "75%"
  }
}
```
Here is my iterator;

``` js
  let rules = ast.ast.stylesheet.rules
  const css = {}

  for (let rule of rules) {
    
    let selector = rule.selectors[0]

    // remove leading dot
    if (selector.charAt(0) === '.') selector = selector.slice(1)

    // remove backslashes
    selector = selector.replace(/\\/g, '')

    let declarations = rule.declarations

    const styles = {}

    for (let declaration of declarations) {
      //camelCase property
      let property = camelCase(declaration.property)
      // convert rem to px values
      let value = remToPx(declaration.value)

      Object.assign(styles, { [property]: value })
    }

    Object.assign(css, { [selector]: styles })
  }


```
